### PR TITLE
feat(widget): Add Picker(with filter) widget

### PR DESCRIPTION
**Pull request by yyanwang**
**Please check if the PR fulfills these requirements**
- [x] The commit(s) message(s) follows our [guidelines](https://github.com/Talend/react-talend-forms/blob/master/CONTRIBUTING.md#commit-message-format)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?**
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)
#24

**What is the new behavior?**
A new widget - Picker with filter is added.

**Does this PR introduce a breaking change?**
- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:
1-This widget is added in story Simple-> TImezone field.
2-And one more question, that I want to add some css for this widget, but I don't know a proper way, seems we only use bootstrap styles here? I can't find a suitable bootstrap class for it - highlight the picker-list item when hover or press UP/DOWN key.

### CODE REVIEW

File: src/widgets/Picker.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+	componentWillReceiveProps(nextProps) {
+		if (nextProps.schema.default !== this.props.schema.default) {
+			const newValue = this.getValue(nextProps.schema.default);
+			this.setState({ value: newValue });
+		}
+	}
+
+	getValue(query) {
+		if (!query) return null;
```
Comment (talend-glorieux) [inline src/widgets/Picker.js]: Please avoid inligned if statements.

File: src/widgets/Picker.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+		const idx = this.props.schema.enum.findIndex(key => key === query);
+		return this.data[idx];
+	}
+
+	filterItems(filter) {
+		if (!filter.trim() === '') return () => true;
```
Comment (talend-glorieux) [inline src/widgets/Picker.js]: Please avoid inligned if statements.

File: src/widgets/Picker.js [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	handleKeyPress(e) {
+		if (e.which === UP_KEY || e.which === DOWN_KEY) {
+			e.preventDefault();
+			let focused = this.state.focused;
+			if (e.which === UP_KEY) {
+				focused -= 1;
+				if (focused < 1) focused = this.state.data.length;
+			} else {
+				focused += 1;
+				if (focused > this.state.data.length) focused = 1;
```
Comment (talend-glorieux) [inline src/widgets/Picker.js]: Please avoid inligned if statements.

File: src/widgets/Picker.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+	handleFilterChange(e) {
+		this.setState({
+			value: e.target.value || '',
+			data: this.data.filter(this.filterItems(e.target.value)),
+			focused: 0,
+		});
+	}
+
+	handleKeyPress(e) {
```
Comment (talend-glorieux) [inline src/widgets/Picker.js]: I'm wondering if a switch case statements calling well named method would improve the logic comprehension. Or maybe just some documentation.

File: src/widgets/Picker.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+			value: this.state.data[index],
+		});
+
+		if (this.props.onChange) {
+			this.props.onChange(this.props.schema.enum[index]);
+		}
+	}
+
+	render() {
+		const value = this.state.value;
```
Comment (talend-glorieux) [inline src/widgets/Picker.js]: That value is only used once. Please remove the assignation.

File: src/widgets/Picker.js [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+		const isOpen = this.state.open;
```
Comment (talend-glorieux) [inline src/widgets/Picker.js]: That value is only used once. Please remove the assignation.

File: src/widgets/Picker.js [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+			>
+				<div>
+					<FormControl
+						onFocus={this.handleFocus}
+						onBlur={this.handleBlur}
+						onChange={this.handleFilterChange}
+						onKeyDown={this.handleKeyPress}
+						value={value}
+						ref={(c) => {
+							this.field = c;
```
Comment (talend-glorieux) [inline src/widgets/Picker.js]: Please inline those arrow functions. `ref={c => this.field = c}`

File: src/widgets/Picker.js [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+						}}
+					/>
+				</div>
+				<ul
+					className="dropdown-menu"
+					ref={(c) => {
+						this.options = c;
+					}}
```
Comment (talend-glorieux) [inline src/widgets/Picker.js]: Please inline those arrow functions. ref={c => this.options = c}

**Comment by jmfrancois** [comment]
please @talend-glorieux could you review the changes.
If it's ok we can update and merge.

**Comment by jmfrancois** [comment]
oh back to wip ?

**Comment by yyanwang** [comment]
@jmfrancois yes, sorry, because I want to apply material-ui design for this widget.

**Comment by yyanwang** [comment]
I am closing this one because currently the SelectWidget already got the feature( filtering with user input). so not so much point in create another.

**Pull request closed without merging**